ledger.rb

```diff
@@ -1,24 +1,156 @@
+# Optional Flags
+
+# --price-db FILE | Use FILE for retrieving stored commodity prices.
+# --file FILE     | Read FILE as a ledger file.
+# --sort VEXPR    | Sort a report using VEXPR.
+
 require 'thor'
+require 'colorize'
 require_relative 'parser'
 
 class Ledger < Thor
+  class_option :file, type: :string, default: 'index.ledger'
+
+  def initialize(*args)
+    super
+    @parser = Parser.new
+    @parser.parse_ledger(options[:file])
+    @balances = {}
+  end
+
   desc "register", "The register command displays all the postings occurring in a single account, line by line."
   def register
-    puts "register command"
+    @parser.parsed_file.each do |transaction|
+      register_title_line(transaction[:date], transaction[:description])
+
+      transaction[:accounts].each do |account|
+        if @balances.include?(account[:currency])
+          @balances[account[:currency]] += account[:amount]
+        else
+          @balances[account[:currency]] = account[:amount]
+        end
+
+        @balances[account[:currency]] = @balances[account[:currency]].round(2)
+        register_line(account[:description], account[:amount], @balances, account[:currency])
+      end
+    end
+
+    register_balance(@balances)
   end
 
   desc "balance", "The balance command reports the current balance of all accounts. "
   def balance
-    parser = Parser.new
-    parsed_info = parser.parse_ledger("index.ledger")
+    transactions = {}
+    transactions_sums = {}
+
+    @parser.parsed_file.each do |transaction|
+      transaction[:accounts].each do |account|
+        if @balances.include?(account[:currency])
+          @balances[account[:currency]] += account[:amount]
+        else
+          @balances[account[:currency]] = account[:amount]
+        end
+
+        @balances[account[:currency]] = @balances[account[:currency]].round(2)
+
+        if transactions_sums.include?(account[:description])
+          transactions_sums[account[:description]] += account[:amount]
+          transactions[account[:description]] = full_action(transactions_sums[account[:description]], account[:currency])
+        else
+          transactions_sums[account[:description]] = account[:amount]
+          transactions[account[:description]] = full_action(account[:amount], account[:currency])
+        end
+      end
+    end
```

Comment:
same as above, you can also make in your class instance variables of `transactions` and `transactions_sums` so you can extract private methods like this:

```ruby
if @transactions_sums.include?(account[:description])
  sum_amount_to_account(account)
else
  add_new_account(account)
end

......

private
def sum_amount_to_account(account)
  @transactions_sums[account[:description]] += account[:amount]
  @transactions[account[:description]] = full_action(@transactions_sums[account[:description]], account[:currency])
end

def add_new_account(account)
  @transactions_sums[account[:description]] = account[:amount]
  @transactions[account[:description]] = full_action(account[:amount], account[:currency])
end
```

Reply:
It's ok to create instance variables even if they are only used in one method and not in all?